File: uva_11151.cpp
# include <iostream>
# include <vector>
# include <string>
using namespace std;
int main(void){
    int num=0;
    string temp;
    cin >> num;
    getline(cin, temp);
    while(num--){
        vector<char> L1, L2;
        int length=0;
        string input_s="";
        getline(cin, input_s);
        length = input_s.length();
        if (length==0){
            cout << "0" << endl;
            continue;
        }
        for (int i=0; i<length; ++i)
            L1.push_back(input_s[i]);
        for (int i=length-1; i>=0; --i)
            L2.push_back(L1[i]);
        int LCS[1001][1001]={0};
        for (int i=1; i<=length; ++i){
            for (int j=1; j<=length; ++j)
                if (L1[j-1]==L2[i-1])
                    LCS[i][j]=LCS[i-1][j-1]+1;
                else 
                    LCS[i][j]=max(LCS[i-1][j], LCS[i][j-1]);
        }
        cout << LCS[length][length] << endl;
    }
    return 0;
}

File: uva_11151.py
num=int(input())
while True:
    if not num:
        break
    num-=1
    l1 = input()
    l2 = l1[::-1]
    length=len(l1)
    if length==0:
        print('0')
        continue
    lcs=[ [0]*1001 for _ in range(1001)]
    for i in range(1, length+1):
        for j in range(1, length+1):
            if l1[j-1]==l2[i-1]:
                lcs[i][j]=lcs[i-1][j-1]+1
            else:
                lcs[i][j]=max(lcs[i-1][j], lcs[i][j-1])

    print( lcs[length][length])
